xkafka/consumer.go

```diff
@@ -111,23 +123,100 @@ func (c *Consumer) Start(ctx context.Context) error {
 				}
 			}
 
+			msg := newMessage(c.name, km)
+
+			err = c.handler.Handle(ctx, msg)
+			if ferr := c.config.errorHandler(err); ferr != nil {
+				errChan <- ferr
+
+				continue
+			}
+
+			err = c.storeMessage(msg)
+			if err != nil {
+				errChan <- err
+
+				continue
+			}
+		}
+	}
+}
+
+func (c *Consumer) runAsync(ctx context.Context) error {
+	st := stream.New().WithMaxGoroutines(c.config.concurrency)
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	for {
+		select {
+		case <-ctx.Done():
+			st.Wait()
+
+			uerr := c.unsubscribe()
+
+			err := context.Cause(ctx)
+			if err != nil && err == context.Canceled {
+				return uerr
+			}
+
+			return errors.Join(err, uerr)
+		default:
+			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
+			if err != nil {
+				if kerr, ok := err.(kafka.Error); ok && kerr.Code() == kafka.ErrTimedOut {
+					continue
+				}
+
+				if ferr := c.config.errorHandler(err); ferr != nil {
+					cancel(ferr)
+
+					continue
+				}
+			}
+
+			msg := newMessage(c.name, km)
+
 			st.Go(func() stream.Callback {
-				c.runHandler(ctx, c.handler, km, errChan)
+				err := c.handler.Handle(ctx, msg)
+				if ferr := c.config.errorHandler(err); ferr != nil {
+					cancel(ferr)
 
-				return func() {}
+					return func() {}
+				}
+
+				return func() {
+					if err := c.storeMessage(msg); err != nil {
+						cancel(err)
+					}
+				}
 			})
 		}
 	}
 }
 
-func (c *Consumer) runHandler(ctx context.Context, handler Handler, km *kafka.Message, errChan chan error) {
-	msg := newMessage(c.name, km)
+func (c *Consumer) storeMessage(msg *Message) error {
+	if msg.Status != Success && msg.Status != Skip {
+		return nil
+	}
 
-	err := handler.Handle(ctx, msg)
+	_, err := c.kafka.StoreOffsets([]kafka.TopicPartition{
+		{
+			Topic:     &msg.Topic,
+			Partition: msg.Partition,
+			Offset:    kafka.Offset(msg.Offset + 1),
+		},
+	})
+	if err != nil {
+		return err
+	}
 
-	if ferr := c.config.errorHandler(err); ferr != nil {
-		errChan <- ferr
+	if c.config.manualCommit {
```

Comment:
This condition may always force a strict commit behaviour. Say I want to commit after every 10th message(or offset), how will this allow that? If that is out of scope for now, is there room for improvement in this area later on?

Reply:
I think we can expose the `Commit` method to allow have their own commit strategy.

So far I have not seen too many use cases of manual commit, hence keeping it simple for now.